sql/2021/performance/web_vitals_by_rank_and_device.sql

```diff
@@ -0,0 +1,166 @@
+#standardSQL
+# Core WebVitals by rank and device
+
+CREATE TEMP FUNCTION IS_GOOD (good FLOAT64, needs_improvement FLOAT64, poor FLOAT64) RETURNS BOOL AS (
+  good / (good + needs_improvement + poor) >= 0.75
+);
+
+CREATE TEMP FUNCTION IS_NI (good FLOAT64, needs_improvement FLOAT64, poor FLOAT64) RETURNS BOOL AS (
+  good / (good + needs_improvement + poor) < 0.75 AND
+  poor / (good + needs_improvement + poor) < 0.25
+);
+
+CREATE TEMP FUNCTION IS_POOR (good FLOAT64, needs_improvement FLOAT64, poor FLOAT64) RETURNS BOOL AS (
+  poor / (good + needs_improvement + poor) >= 0.25
+);
+
+CREATE TEMP FUNCTION IS_NON_ZERO (good FLOAT64, needs_improvement FLOAT64, poor FLOAT64) RETURNS BOOL AS (
+  good + needs_improvement + poor > 0
+);
+
+WITH
+base AS (
+  SELECT
+    date,
+    origin,
+    device,
+    rank,
+
+    fast_fid,
+    avg_fid,
+    slow_fid,
+
+    fast_lcp,
+    avg_lcp,
+    slow_lcp,
+
+    small_cls,
+    medium_cls,
+    large_cls,
+
+    fast_fcp,
+    avg_fcp,
+    slow_fcp,
+
+    fast_ttfb,
+    avg_ttfb,
+    slow_ttfb
+
+  FROM
+    `chrome-ux-report.materialized.device_summary`
+  WHERE
+    device IN ('desktop', 'phone') AND
+    date IN ('2021-07-01')
+)
+
+SELECT
+  date,
+  device,
+  CASE
+    WHEN rank_grouping = 10000000 THEN 'all'
+    ELSE CAST(rank_grouping AS STRING)
+  END AS ranking,
+
+  COUNT(DISTINCT origin) AS total_origins,
+
+  SAFE_DIVIDE(
+    COUNT(DISTINCT IF(
+        IS_GOOD(fast_fid, avg_fid, slow_fid) AND
+        IS_GOOD(fast_lcp, avg_lcp, slow_lcp) AND
+        IS_GOOD(small_cls, medium_cls, large_cls), origin, NULL)),
+    COUNT(DISTINCT IF(
+        IS_NON_ZERO(fast_fid, avg_fid, slow_fid) AND
```

Comment:
Missed this initially. Since many sites have no FID data, in our CWV evaluations we should consider it optional.

See the CWV Tech Report queries for an example of calculating the [numerator](https://github.com/HTTPArchive/cwv-tech-report/blob/50705bab1e712218c679e86ffac9995fb650452e/sql/monthly.sql#L56-L58) and [denominator/pct](https://github.com/HTTPArchive/cwv-tech-report/blob/50705bab1e712218c679e86ffac9995fb650452e/sql/monthly.sql#L121)